### DISCUSSION TITLE
Having trouble understanding "Left recurrsion"

### DISCUSSION THREAD

**Discussion by Thesecondbestname** [pest-parser/pest#786] [Q&A] [answered]
This code 
`call                  = { (ident ~ ("(" ~ fn_args? ~ ")")?) | method_call}
method_call   = { expression ~ "." ~ ident ~ ( "(" ~ fn_args? ~ ")" )? }
expression     = !{ (math_expr  | bool_expr | value | block | if_statement | call) ~ lookahead? }
`
Is "Left recursive".  
> rule method_call is left-recursive (method_call -> call -> method_call)

I don't quite get what this means. While the input isn't infinite this should work fine since there are a lot of ways to break out of this loop. I want to be able to call methods on expressions that **may or may not** contain another expression call or method.
I am really, really sorry for my frustrated tone. I've been trying to rack my brain around for sometime and have the sneaking suspicion that it's something really quite simple I'm way too tired to notice. Thank you for your time.

**Comment by zacryol** [comment, marked as answer]
Smaller example:

```
S = { (S ~ "+") | "a" }
```

On paper, this grammar appears to be able to parse `"a"` followed by any number of `"+"`; but with PEG rules, that can't actually happen.

PEGs are eager and top down. When a rule is encountered, pest will try to match that rule entirely before going forward. With the above grammar, when you try to parse an `S` from some input, pest will try to complete an inner `S` first, but to parse that `S`, it will need to parse an `S` first, but to parse *that* `S`, it will need to parse an `S` first.

**No input gets consumed** before you end up with trying to parse the same rule, therefore no progress can be made.

"Left recursion" means that a right hand option can start with the same left hand nonterminal (perhaps with a few extra steps).

In your specific grammar, when parsing a `call`, pest can try to start with a `method_call`, which can start with an `expression`, which can start with a `call`, and then you end up with nothing but a larger call stack.

**Reply by Thesecondbestname** [reply]
Thanks a LOT! Once again I am sorry for my tone. It was very late. I actually remember having asked this sometime back elsewhere. Thanks you for your patience regardless!
